internal/pkg/client/patch.go

```diff
@@ -0,0 +1,207 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/apimachinery/pkg/util/jsonmergepatch"
+	"k8s.io/apimachinery/pkg/util/mergepatch"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+)
+
+// This file has dependency on kubectl
+// Ideally we would need to figure out a way to deprecate this dependency
+
+var metadataAccessor = meta.NewAccessor()
+
+// GetClientSideApplyPatch updates a resource using the same approach as running `kubectl apply`.
+// The implementation here has been mostly extracted from the apply command: k8s.io/kubernetes/pkg/kubectl/cmd/apply.go
+func GetClientSideApplyPatch(current, desired runtime.Object) (Patch, error) {
+	var patch Patch
+	// Serialize the current configuration of the object.
+	currentb, err := runtime.Encode(unstructured.UnstructuredJSONScheme, current)
+	if err != nil {
+		return patch, fmt.Errorf("could not serialize current %v", current)
+	}
+	// Retrieve the last applied configuration of the object from the annotation.
+	last, err := getLastApplied(current)
+	if err != nil {
+		return patch, fmt.Errorf("could not retrieve original configuration from %v", desired)
+	}
+	fmt.Printf("lastAppliedPatch: %s\n", string(last))
+
+	// Serialize the modified configuration of the object, populating the last applied annotation as well.
+	modified, err := GetModifiedConfiguration(desired, true, unstructured.UnstructuredJSONScheme)
+	if err != nil {
+		return patch, fmt.Errorf("could not serialize intended configuration from %v", desired)
+	}
+	gvk := desired.GetObjectKind().GroupVersionKind()
+
+	resourceDescription := fmt.Sprintf("%s/%s", gvk, "TODO")
+	versionedObject, err := Scheme.New(gvk)
+	_, unversioned := Scheme.IsUnversioned(desired)
+	switch {
+	case runtime.IsNotRegisteredError(err) || unversioned:
+		preconditions := []mergepatch.PreconditionFunc{
+			mergepatch.RequireKeyUnchanged("apiVersion"),
+			mergepatch.RequireKeyUnchanged("kind"),
+			mergepatch.RequireMetadataKeyUnchanged("name"),
+		}
+		patch.Data, err = jsonmergepatch.CreateThreeWayJSONMergePatch(last, modified, currentb, preconditions...)
+		patch.Type = types.MergePatchType
+		if err != nil {
+			if mergepatch.IsPreconditionFailed(err) {
+				return patch, fmt.Errorf("at least one of apiVersion, kind and name was changed for %s", resourceDescription)
+			}
+			return patch, err
+		}
+	case err != nil:
+		return patch, err
+	case err == nil:
+		// Compute a three way strategic merge patch to send to server.
+		patch.Type = types.StrategicMergePatchType
+		lookupPatchMeta, err := strategicpatch.NewPatchMetaFromStruct(versionedObject)
+		if err != nil {
+			return patch, err
+		}
+		patch.Data, err = strategicpatch.CreateThreeWayMergePatch(last, modified, currentb, lookupPatchMeta, true)
+		if err != nil {
+			return patch, err
+		}
+	}
+	return patch, nil
+}
+
+// GetMergePatch - generate merge patch from original and modified objects
+func GetMergePatch(original, modified runtime.Object) (Patch, error) {
```

Comment:
Where is this used?

Reply:
Not used currently.
Thinking of a use-case where 2 way merge patch without lastappliedannotation.
Will be useful to port over to controller runtime.